test/server/config_validation/xds_fuzz.cc

```diff
@@ -184,77 +186,244 @@ void XdsFuzzTest::replay() {
+      } else {
+        updateRoute(routes_, {}, {});
       }
       break;
     }
     default:
       break;
     }
+
+    // TODO(samflattery): makeSingleRequest here?
+    version_++;
   }
 
+  verifyState();
   close();
 }
 
+void XdsFuzzTest::drainListener(const std::string& name) {
+  // ensure that the listener drains correctly by checking that it is still draining halfway through
+  ENVOY_LOG_MISC(info, "Draining {}", name);
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining()); */
+  /* timeSystem().advanceTimeWait(std::chrono::milliseconds(300000)); */
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining()); */
+  /* timeSystem().advanceTimeWait(std::chrono::milliseconds(300001)); */
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining() - 1); */
+  verifier_.drainedListener(name);
+}
+
+void XdsFuzzTest::verifyListeners() {
+  ENVOY_LOG_MISC(info, "Verifying listeners");
+  const auto& listeners = verifier_.listeners();
+  ENVOY_LOG_MISC(info, "There are {} listeners in listeners_", listeners.size());
+  envoy::admin::v3::ListenersConfigDump listener_dump = getListenersConfigDump();
+  ENVOY_LOG_MISC(info, "{}", listener_dump.DebugString());
+  for (auto& listener_rep : listeners) {
+    ENVOY_LOG_MISC(info, "Verifying {} with state {}", listener_rep.listener.name(),
+                   listener_rep.state);
+    bool found = false;
+    for (auto& dump_listener : listener_dump.dynamic_listeners()) {
+      if (dump_listener.name() == listener_rep.listener.name()) {
+        ENVOY_LOG_MISC(info, "Found a matching listener in config dump");
+        ENVOY_LOG_MISC(info, "drain: {}, warm {}, active {}", dump_listener.has_draining_state(),
+                       dump_listener.has_warming_state(), dump_listener.has_active_state());
+        switch (listener_rep.state) {
+        case XdsVerifier::DRAINING:
+          if (dump_listener.has_draining_state()) {
+            ENVOY_LOG_MISC(info, "{} is draining", listener_rep.listener.name());
+            drainListener(listener_rep.listener.name());
+            found = true;
+          }
+          break;
+        case XdsVerifier::WARMING:
+          if (dump_listener.has_warming_state()) {
+            ENVOY_LOG_MISC(info, "{} is warming", listener_rep.listener.name());
+            found = true;
+          }
+          break;
+        case XdsVerifier::ACTIVE:
+          if (dump_listener.has_active_state()) {
+            ENVOY_LOG_MISC(info, "{} is active", listener_rep.listener.name());
+            found = true;
+          }
+          break;
+        default:
+          NOT_REACHED_GCOVR_EXCL_LINE;
+        }
+      }
+    }
+    if (!found) {
+      ENVOY_LOG_MISC(info, "Expected to find listener {} in config dump",
```

Comment:
I see. So at a high-level, we keep track of listenres in the verifier, and then go through the list of listeners in the listener dump, find it's representative in the verifier, and compare their states? 'found' is a representation of "is that listener in the verifier AND does it's state match?"

Reply:
Right.  For each listener in the verifier, we check to see that there is a matching listener in the config dump, i.e. listener with same name and state